File: src/lab5C/ffs_ref.c
#include <stdio.h>
#include <stdlib.h>

typedef unsigned int uint32_t;

uint32_t ffs_ref(uint32_t word) {
   int i = 0;
	int cnt = 0;
   if (!word) return 0;
   for (cnt = 0; cnt < 32; cnt++)
      if (((1 << i++) & word) != 0) return i;
   return 0; /* notreached */
}

int main (int argc, char **argv) {
   if (argc != 2) {
      printf("Usage: %s <number>\n", argv[0]);
      exit(0);
   }
   uint32_t n = atol(argv[1]);
   printf("ffs_ref: first 1 at %d\n", ffs_ref(n));
}

File: src/lab5C/ffs_ref.java
// compile with javac -g ffs_ref.java (-g generates all debug information)
public class ffs_ref {
   static int ffs_ref(int word) {
      int i = 0;
      if (word == 0) return 0;
      for(int cnt = 0; cnt < 32; cnt++) if (((1 << i++) & word) != 0) return i;
      return 0;
   }
   
   public static void main (String args[]) {
      int k = Integer.parseInt(args[0]);
      System.out.println("First 1 is at "+ffs_ref(k));
   }
}
